crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -111,6 +132,43 @@ struct TypeInferenceBuilder<'db> {
     types: TypeInference<'db>,
 }
 
+/// A builder is used by creating it with `::new`, and then calling `.finish()` on it, which
+/// returns the resulting [`TypeInference`].
+///
+/// There are a few different kinds of methods in the type inference builder, and the naming
+/// distinctions are a bit subtle.
+///
+/// The `finish` method calls `infer_region`, which delegates to one of `infer_region_scope`,
+/// `infer_region_definition`, or `infer_region_expression`, depending which kind of
+/// [`InferenceRegion`] we are inferring types for.
+///
+/// Scope inference starts with the scope body, walking all statements and expressions and
+/// recording the types of each expression in the [`TypeInference`] result. Most of the methods
+/// here (with names like `infer_*_statement` or `infer_*_expression` or some other node kind) take
+/// a single AST node and are called as part of this AST visit.
+///
+/// When the visit encounters a node which creates a [`Definition`], we look up the definition in
+/// the semantic index and call the `infer_definition_types` query on it, which creates another
```

Comment:
```suggestion
/// the semantic index and call the [`infer_definition_types`] query on it, which creates another
```

Reply:
FWIW rustdoc requires this to be `[\`infer_definition_types()\`]` (with the parens), because Salsa's macros create both a struct and a function by this name, and the link must disambiguate.